blazar/tests/enforcement/test_enforcement.py

```diff
@@ -0,0 +1,209 @@
+
+    def check_create(self, context, lease_values):
+        pass
+
+    def check_update(self, context, current_lease_values, new_lease_values):
+        pass
+
+    def on_end(self, context, lease_values):
+        pass
+
+
+@ddt.ddt
+class EnforcementTestCase(tests.TestCase):
+    def setUp(self):
+        super(EnforcementTestCase, self).setUp()
+
+        self.cfg = cfg
+        self.region = 'RegionOne'
+        self.all_filters = ['FakeFilter']
+        self.filters = filters
+
+        self.enforcement = enforcement.UsageEnforcement()
+        self.filters.FakeFilter = FakeFilter
+        cfg.CONF.set_override(
+            'enabled_filters', self.all_filters, group='enforcement')
+        cfg.CONF.set_override('os_region_name', self.region)
+
+        self.enforcement.load_filters()
+        self.fake_service_catalog = [
+            dict(
+                type='identity', endpoints=[
+                    dict(
+                        interface='public', region=self.region,
+                        url='https://fakeauth.com')
+                ]
+            )
+        ]
+
+        self.ctx = context.BlazarContext(
+            user_id=111, project_id=222,
+            service_catalog=self.fake_service_catalog)
+        self.set_context(self.ctx)
+
+        self.fake_host_id = '1'
+        self.fake_host = {
+            'id': self.fake_host_id,
+            'hypervisor_hostname': 'hypvsr1',
+            'service_name': 'compute1',
+            'vcpus': 4,
+            'cpu_info': 'foo',
+            'hypervisor_type': 'xen',
+            'hypervisor_version': 1,
+            'memory_mb': 8192,
+            'local_gb': 10,
+        }
+
+    def tearDown(self):
+        super(EnforcementTestCase, self).tearDown()
+
+    def test_load_filters(self):
+        print(self.enforcement.enabled_filters)
+        self.assertEqual(len(self.enforcement.enabled_filters), 1)
+
+        fake_filter = self.enforcement.enabled_filters[0]
+
+        self.assertEqual(fake_filter.conf.enforcement.fake_opt, 1)
+
+    def test_format_context(self):
+
+        formatted_context = self.enforcement.format_context(context.current())
+
+        expected = dict(user_id=111, project_id=222, region_name=self.region,
+                        auth_url='https://fakeauth.com')
+
+        self.assertDictEqual(expected, formatted_context)
+
+    def test_check_create(self):
+        allocation_candidates = {'virtual:instance': [get_fake_host('1')]}
+        lease_values = get_fake_lease()
+        reservations = list(lease_values['reservations'])
+
+        del lease_values['reservations']
+        ctx = context.current()
+
+        check_create = self.patch(self.enforcement.enabled_filters[0],
+                                  'check_create')
+
+        self.enforcement.check_create(ctx, lease_values, reservations,
+                                      allocation_candidates)
+
+        formatted_lease = self.enforcement.format_lease(
```

Comment:
Looking at this test, I think the most valuable thing to test are these `format_lease`and `format_context` functions. This test you have checks that some function is called with some signature. That's OK, but it's not of very high value, relative to checking if the formatting of those arguments is sane.

Another valuable test would be to mock the filter's check_create and have it raise an exception, and then make sure that the lease doesn't get created.

Reply:
So I added a test for lease_format. I also added checks in each main check_create, check_update, and on_end test.

For mocking check_create, I added tests to service that assert that the UsageEnforcement classes check_create function raises an error and then make sure the lease is not created. In the tests enforcement, I test that if a filter raises an exception that check_create will also raise that exception. I think this the way to go since we have to mock enforcement in the service tests, we can't easily mock the filters deeper down the stack.